llvm/lib/CAS/BuiltinCAS.cpp

```diff
@@ -353,784 +287,2929 @@ Error BuiltinCAS::printCASID(raw_ostream &OS, CASID ID) {
+  BuiltinBlobHasher<HasherT> Hasher;
+  HashType ComputedHash = Hasher.hash(Data);
+  if (!isAligned(Align(PageSize), Data.size()) && Data.end()[0] == 0)
+    return createBlobFromNullTerminatedRegion(ComputedHash, std::move(Map));
+  return createBlobImpl(ComputedHash, Data);
+}
 
-static void printTree(raw_ostream &OS, ArrayRef<NamedTreeEntry> SortedEntries,
-                      SmallVectorImpl<CASID> &References) {
-  // Hash input has format:
-  //
-  //  input := header unnamed-entry* name*
-  //  header := 'cas-tree' num-entries
-  //  unnamed-entry := kind name-offset
-  //  name := name-string NULL
-  OS << CASTreeMagic;
-  support::endian::Writer EW(OS, support::endianness::little);
-  uint64_t StringsOffset = 0;
-  for (const NamedTreeEntry &Entry : SortedEntries) {
-    assert(Entry.getID().getHash().size() == NumHashBytes);
-    References.push_back(Entry.getID());
-
-    uint64_t Kind = 0;
-    switch (Entry.getKind()) {
-    case TreeEntry::Regular:
-      Kind = EntryKindIDForRegular;
-      break;
-    case TreeEntry::Tree:
-      Kind = EntryKindIDForTree;
-      break;
-    case TreeEntry::Executable:
-      Kind = EntryKindIDForExecutable;
-      break;
-    case TreeEntry::Symlink:
-      Kind = EntryKindIDForSymlink;
-      break;
-    }
-    assert(Kind < (1u << 8) && "Kind needs to be 8 bits");
+Expected<TreeRef> BuiltinCAS::createTree(ArrayRef<NamedTreeEntry> Entries) {
+  // Ensure a stable order for tree entries and ignore name collisions.
+  SmallVector<NamedTreeEntry> Sorted(Entries.begin(), Entries.end());
+  std::stable_sort(Sorted.begin(), Sorted.end());
+  Sorted.erase(std::unique(Sorted.begin(), Sorted.end()), Sorted.end());
 
-    assert(StringsOffset < (1ull << 56) && "Absurdly large string?");
-    EW.write(StringsOffset | (Kind << 56));
-    StringsOffset +=
-        alignTo(Entry.getName().size() + 1, Align(4)) + sizeof(uint32_t);
-  }
-  for (const NamedTreeEntry &Entry : SortedEntries) {
-    EW.write(uint32_t(Entry.getName().size()));
-    OS << Entry.getName();
-    for (int I = 0, E = alignTo(Entry.getName().size() + 1, Align(4)) -
-                        Entry.getName().size();
-         I != E; ++I)
-      OS.write(0);
-  }
+  BuiltinTreeHasher<HasherT> Hasher;
+  Hasher.start(Sorted.size());
+  for (const NamedTreeEntry &E : Sorted)
+    Hasher.updateEntry(E.getID().getHash(), E.getName(), E.getKind());
+  return createTreeImpl(Hasher.finish(), Sorted);
 }
 
-CASID BuiltinCAS::ObjectContentReference::getReference(size_t I) const {
-  assert(I < getNumReferences() && "Reference out of range");
-  return CASID(
-      bufferAsRawHash(ReferenceBlock.substr(I * NumHashBytes, NumHashBytes)));
+Expected<NodeRef> BuiltinCAS::createNode(ArrayRef<CASID> References,
+                                         ArrayRef<char> Data) {
+  BuiltinNodeHasher<HasherT> Hasher;
+  Hasher.start(References.size());
+  for (const CASID &ID : References)
+    Hasher.updateRef(ID.getHash());
```

Comment:
`ID.getHash()` returns the CAS internal representation right? Should this be based on something more stable like the hash value with the actual hash algorithm?

For example, if you have InMemoryCAS and OnDiskCAS, they can use the same hash function but different internal representation. If you use the implementation specific value, the identical node is going to have different ID in those two CAS so you can't overlay or simple transfer between them.

Reply:
Never mind, I read the comment wrong.